### ISSUE TITLE
Extract hidden state representation

### ISSUE THREAD

**Issue by macabdul9** [nyu-mll/jiant#1251]
I wanted to extract hidden state representation of nth layer from the model we're loading. I went through the `wrap_jiant_forward` function which returns union of 3 things based on the `keys` value. How can I change it to return only hidden state representation of nth layer, so that instead of the logits output I send those hidden state representations to the `complex_backpropagate` function?

**Comment by bhavitvyamalik** [comment]
Hi @jeswan, 
Can you help with this too?

**Comment by jeswan** [comment]
The following utility function is suitable for your purpose:
https://github.com/nyu-mll/jiant/blob/32fec3659cae03e3d68a4eda0e20a6319f20c2ae/jiant/utils/transformer_utils.py#L6-L32
Here is an example how to use the function to extract hidden states of the nth layer:
https://github.com/nyu-mll/jiant/blob/04bbb39f97d1074b03b55237c4dfa363932d26c7/jiant/proj/main/modeling/taskmodels.py#L237-L241

**Comment by bhavitvyamalik** [comment]
1) I was looking into `EmbeddingModel` class and I think this is what I need. How can I choose my `TaskTypes` (saw that in `core.py`) in this [notebook](https://github.com/nyu-mll/jiant/blob/master/examples/notebooks/jiant_Multi_Task_Example.ipynb) as I want embeddings from model. 

From what I understood, this is what I should I add in model config inorder to get output from layer 0 and feed it back into backprop function?
```
{
    "model_type": "bert-base-uncased",
    "model_path": "./models/bert-base-uncased/model/bert-base-uncased.p",
    "model_config_path": "./models/bert-base-uncased/model/bert-base-uncased.json",
    "model_tokenizer_path": "./models/bert-base-uncased/tokenizer",
    "layer": 0
  }
```
If I don't pass `pool` parameter, what's the default pooling done?

2) Should I freeze the model manually in `export_model.py` before you save its weights or is there a parameter for the same?

**Comment by zphang** [comment]
Hi @bhavitvyamalik,

To clarify, the `EmbeddingModel` returns a single vector representation (e.g. 768-dim) for the whole input, extracted from one of the layers. Currently, it is used for tasks like Tatoeba and BUCC2018, which are retrieval tasks evaluated based on their  nearest neighbors in the embedding space.

More specifically answering your question:

1. The configuration for the pooler-type and layer would be within the taskmodel configuration. This is found in the "runconfig" (a.k.a. `jiant_task_container`). You'll want it to be set at the key `runconfig["taskmodels_config"][TASKMODEL_NAME]["layer"]` and `runconfig["taskmodels_config"][TASKMODEL_NAME]["pooler_type"]`, where `TASKMODEL_NAME` is generally the task name.

2. The model weights are not frozen in regular operation of the `main/runscript.py`. There is currently no flag to do so, but you can modify the code to freeze weights by setting `p.requires_grad_(False)` for all parameters you want to freeze. You can insert code here:

https://github.com/nyu-mll/jiant/blob/ce6249582d3eb7dbc37c577a6990f3d398b77f34/jiant/proj/main/runscript.py#L96

Something along the lines of

```python
for n, p in jiant_model.named_parameters():
    if SOME_CONDITION:
        p.requires_grad_(False)
```

**Comment by jeswan** [comment]
Closed due to inactivity.
